hibernate-osgi/src/main/java/org/hibernate/osgi/OsgiClassLoader.java

```diff
@@ -8,33 +8,41 @@
 
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Enumeration;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.NoSuchElementException;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 import org.hibernate.service.spi.Stoppable;
 import org.osgi.framework.Bundle;
 
 /**
  * Custom OSGI ClassLoader helper which knows all the "interesting"
  * class loaders and bundles.  Encapsulates the OSGi related CL capabilities.
- * 
+ *
  * @author Brett Meyer
  * @author Tim Ward
  */
 public class OsgiClassLoader extends ClassLoader implements Stoppable {
 	// Leave these as Sets -- addClassLoader or addBundle may be called more
 	// than once if a SF or EMF is closed and re-created.
-	private Set<ClassLoader> classLoaders = new LinkedHashSet<ClassLoader>();
-	private Set<Bundle> bundles = new LinkedHashSet<Bundle>();
+	// HHH-12553: must be thread-safe. Concurrent impl. would be best, but we have to retain insertion-order.
+	private Set<ClassLoader> classLoaders = Collections.synchronizedSet(new LinkedHashSet<ClassLoader>());
+	private Set<Bundle> bundles = Collections.synchronizedSet(new LinkedHashSet<Bundle>());
+
+	private ConcurrentMap<String, Class<?>> classCache = new ConcurrentHashMap<String, Class<?>>();
```

Comment:
It looks like some caching is introduced. Was that really necessary? Can someone reassure me about leaks?

Also, it would help to have some comments about synchronization expecations. Why making these two use `ConcurrentHashMap` and then extensively `synchronize` on `bundles` and/or `classLoaders` ?

Reply:
@Sanne Thanks for looking into it.
The registerAsParallelCapable you commented on above is not specific to OSGi, but to ClassLoaders in general.
See my comments in HHH-12553 for some more details.
The caching was not introduced by this change, but was already there and I retained it of course.
There is no synchronization on the concurrent maps, only on the not-thread-safe maps, see comment in code: "must be thread-safe. Concurrent impl. would be best, but we have to retain insertion-order."

Reply:
@Sanne, no new caching here -- @MrEasy is purely changing the data types.  This is where we hold all of the discovered OSGi bundles/CLs for use in our aggregated CL.  These need to be discovered during bootstrapping and held permanently for the ORM bundles.